### PR TITLE
Add ensure logged in to all auth protected pages and choice of active role

### PR DESCRIPTION
**Pull request by jamestjw**
Add ensure logged in to all pages that should ohly be viewed by logged in users. 

Allow the toggling of roles

### CODE REVIEW

File: public/discussion.php [hunk 1]
```diff
@@ -1,8 +1,8 @@
+<?php require_once (dirname(__FILE__)."/../modules/ensure_logged_in.php"); ?>
+
 <?php
 require_once "../modules/models/discussion.php";
-require_once "../common.php";
-
-ensure_logged_in();
+require_once (dirname(__FILE__)."/./modules/ensure_logged_in.php");
```
Comment (chris7neves) [inline public/discussion.php:5]: Might be wrong, but is this included twice in the same file? Is there a reason for that?

Comment (jamestjw) [reply public/discussion.php:5]: Whoops you're right

File: public/templates/footer.php [hunk 2]
```diff
@@ -1,2 +1,3 @@
+<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous"></script>
```
Comment (chris7neves) [inline public/templates/footer.php:1]: were using bootstrap?

File: public/templates/header.php [hunk 3]
```diff
@@ -1,13 +1,21 @@
...
+
 <!DOCTYPE html>
 <html lang="en">
   <head>
     <meta charset="utf-8" />
     <meta http-equiv="x-ua-compatible" content="ie=edge" />
     <meta name="viewport" content="width=device-width, initial-scale=1" />
 
-    <script type="text/javascript" src="https://code.jquery.com/jquery-3.6.0.slim.min.js" integrity="sha256-u7e5khyithlIdTpu22PHhENmPcRdFiHRjhAuHcs05RI=" crossorigin="anonymous"></script>
+    <script src="https://code.jquery.com/jquery-3.6.0.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>
```
Comment (chris7neves) [inline public/templates/header.php:16]: are we still using jquery?

File: public/templates/footer.php (hunk 2 above)
Comment (jamestjw) [reply public/templates/footer.php:1]: erm.. yeah i'll remove it i guess

File: public/templates/header.php (hunk 3 above)
Comment (jamestjw) [reply public/templates/header.php:16]: yeah i'll ditch this too, i can do the same thing without it i think

**Comment by jamestjw** [comment]
@chris7neves Okay I changed it all to use vanilla JS and css 😂  the code looks like shit now

oh and do try running it just in case I missed something :) I feel less confident when I have don't use jQuery

File: modules/models/user.php [hunk 4]
```diff
@@ -9,12 +9,30 @@ class User extends Record
...
+        }
+
+        if ($this->is_instructor) {
+            array_push($res, "instructor");
+        }
+
+        if (isset($this->student_id)) {
+            array_push($res, "student");
+        }
+
```
Comment (drouba) [inline modules/models/user.php:35]: Should we add TA here as well?

File: public/css/header.css [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  text-align: left;
+}
+
+.dropdown-content a:hover {background-color: #f1f1f1;}
+
+.dropdown:hover .dropdown-content {
+  display: block;
+}
+
+.dropdown-content a.active {background-color: #00ff00;}
```
Comment (drouba) [inline public/css/header.css:60]: Why this screamy green, please change to 47D380

**Review by drouba — COMMENTED** [review]
why as admin is the only available role Admin? 
Admin should have access to all roles no?

**Comment by jamestjw** [comment]
> why as admin is the only available role Admin?
> Admin should have access to all roles no?

ah yes, thats a good point 🤔 ok I'll implement that

File: modules/models/user.php (hunk 4 above)
Comment (jamestjw) [reply modules/models/user.php:35]: @drouba do we have a way of telling if someone is a TA? I don't see it

File: public/css/header.css (hunk 5 above)
Comment (jamestjw) [reply public/css/header.css:60]: ok sure 🆗

**Pull request merged**